File: Unit 2-Data Structure/Queue/As Array/Queue.py
class AQueue:
    MAX_SIZE = 100

    def __init__(self):
        self.queue = [None] * self.MAX_SIZE
        self.front = self.rear = -1

    def enqueue(self, value):
        if self.is_full():
            print("Queue Overflow")
        else:
            if self.is_empty():
                self.front = self.rear = 0
            else:
                self.rear = (self.rear + 1) % self.MAX_SIZE
            self.queue[self.rear] = value

    def dequeue(self):
        if self.is_empty():
            print("Queue Underflow")
            return None
        else:
            item = self.queue[self.front]
            if self.front == self.rear:
                self.front = self.rear = -1
            else:
                self.front = (self.front + 1) % self.MAX_SIZE
            return item

    def is_empty(self):
        return self.front == -1 and self.rear == -1

    def is_full(self):
        return (self.rear + 1) % self.MAX_SIZE == self.front


if __name__ == "__main__":
    my_queue = AQueue()

    for i in range(1, 10):
        my_queue.enqueue(i)

    print("Dequeue:", my_queue.dequeue())
    print("Is the queue empty?", "Yes" if my_queue.is_empty() else "No")

File: Unit 2-Data Structure/Queue/As Array/Queue.cpp
#include <iostream>

using namespace std;

template <typename T>
class AQueue
{
private:
    static const int MAX_SIZE = 100;
    T queue[MAX_SIZE];
    int front, rear;

public:
    AQueue() : front(-1), rear(-1) {}

    void enqueue(T value)
    {
        if (isFull()) {
            cout << "Queue Overflow" << endl;
        } else {
            if (isEmpty()) {
                front = rear = 0;
            } else {
                rear = (rear + 1) % MAX_SIZE;
            }
            queue[rear] = value;
        }
    }

    T dequeue()
    {
        if (isEmpty())
        {
            cout << "Queue Underflow" << endl;
            return T();
        } 
        else
        {
            T item = queue[front];
            if (front == rear)
            {
                front = rear = -1;
            }
            else
            {
                front = (front + 1) % MAX_SIZE;
            }
            return item;
        }
    }

    bool isEmpty()
    {
        return front == -1 && rear == -1;
    }

    bool isFull()
    {
        return (rear + 1) % MAX_SIZE == front;
    }
};

int main()
{
    AQueue<int> myQueue;

    for (int i = 1; i < 10; i++)
    {
        myQueue.enqueue(i);
    }

    cout << "Dequeue: " << myQueue.dequeue() << endl;
    cout << "Is the queue empty? " << (myQueue.isEmpty() ? "Yes" : "No") << endl;

    return 0;
}
